src/analysis/normed_space/lp_space.lean

```diff
@@ -0,0 +1,515 @@
+/-
+Copyright (c) 2021 Heather Macbeth. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Heather Macbeth
+-/
+import analysis.mean_inequalities
+import analysis.mean_inequalities_pow
+import topology.algebra.ordered.liminf_limsup
+
+/-!
+# ℓp space
+
+This file describes properties of elements `f` of a pi-type `Π i, E i` with finite "norm",
+defined for `p:ℝ≥0∞` as the size of the support of `f` if `p=0`, `(∑' a, ∥f a∥^p) ^ (1/p)` for
+`0 < p < ∞` and `⨆ a, ∥f a∥` for `p=∞`.
+
+The Prop-valued `mem_ℓp f p` states that a function `f : Π i, E i` has finite norm according
+to the above definition; that is, `f` has finite support if `p = 0`, `summable (λ a, ∥f a∥^p)` if
+`0 < p < ∞`, and `bdd_above (norm '' (set.range f))` if `p = ∞`.
+
+The space `lp E p` is the subtype of elements of `Π i : α, E i` which satisfy `mem_ℓp f p`. For
+`1 ≤ p`, the "norm" is genuinely a norm and (TODO) `lp` is a complete metric space.
+
+## Main definitions
+
+* `mem_ℓp f p` : property that the function `f` satisfies, as appropriate, `f` finitely supported
+  if `p = 0`, `summable (λ a, ∥f a∥^p)` if `0 < p < ∞`, and `bdd_above (norm '' (set.range f))` if
+  `p = ∞`
+* `lp E p` : elements of `Π i : α, E i` such that `mem_ℓp f p`. Defined as an `add_subgroup` of
+  `Π i : α, E i`.
+
+## Implementation
+
+Since `lp` is defined as an `add_subgroup`, dot notation does not work. Use `lp.norm_neg f` to
+say that `∥-f∥ = ∥f∥`, instead of the non-working `f.norm_neg`.
+
+## TODO
+
+* Hölder's inequality
+* Completeness of `lp`
+* Equivalence with `pi_Lp`, for `α` finite
+* Equivalence with `measure_theory.Lp`, for `f : α → E` (i.e., functions rather than pi-types) and
+  the counting measure on `α`
+* Equivalence with `bounded_continuous_function`, for `f : α → E` (i.e., functions rather than
+  pi-types) and `p = ∞`, and the discrete topology on `α`
+
+-/
+
+noncomputable theory
+open_locale nnreal ennreal big_operators
+
+variables {α : Type*} {E : α → Type*} {p q : ℝ≥0∞} [Π i, normed_group (E i)]
+
+/-!
+### `mem_ℓp` predicate
+
+-/
+
+/-- The property that `f : Π i : α, E i`
+* is finitely supported, if `p = 0`, or
+* admits an upper bound for `set.range (λ i, ∥f i∥)`, if `p = ∞`, or
+* has the series `∑' i, ∥f i∥ ^ p` be summable, if `0 < p < ∞`. -/
+def mem_ℓp (f : Π i, E i) (p : ℝ≥0∞) : Prop :=
+if p = 0 then (set.finite {i | f i ≠ 0}) else
+  (if p = ∞ then bdd_above (set.range (λ i, ∥f i∥)) else summable (λ i, ∥f i∥ ^ p.to_real))
+
+lemma mem_ℓp_zero {f : Π i, E i} (hf : set.finite {i | f i ≠ 0}) : mem_ℓp f 0 :=
```

Comment:
Is it worth upgrading this to an iff:
```suggestion
lemma mem_ℓp_zero_iff {f : Π i, E i} : set.finite {i | f i ≠ 0} ↔ mem_ℓp f 0 :=
by simp [mem_ℓp]
```

Reply:
Oh I now see you also provide `mem_ℓp.finite_dsupport` below for the other directions to enable dot notation.

In this case, we should keep `mem_ℓp_zero` as you have it but maybe also provide `mem_ℓp_zero_iff` (and likewise for `p > 0` cases).

Reply:
Done, for all three!